AdyenCard/Utilities/CardType/BinInfoProvider.swift

```diff
@@ -46,59 +48,58 @@ internal final class BinInfoProvider: AnyBinInfoProvider {
     ///   - bin: Card's BIN number. If longer than `minBinLength` - calls API, otherwise check local Regex.
     ///   - supportedTypes: Card brands supported by the merchant.
     ///   - completion:  Callback to notify about results.
-    internal func provideInfo(for bin: String, supportedTypes: [CardType], completion: @escaping (BinLookupResponse) -> Void) {
+    internal func provide(for bin: String, supportedTypes: [CardType], completion: @escaping (BinLookupResponse) -> Void) {
+        let fallback: () -> Void = {
+            self.fallbackCardTypeProvider.provide(for: bin,
+                                                  supportedTypes: supportedTypes,
+                                                  completion: completion)
+        }
+
         guard bin.count >= BinInfoProvider.minBinLength else {
-            return fallbackCardTypeProvider.provideInfo(for: bin,
-                                                        supportedTypes: supportedTypes,
-                                                        completion: completion)
+            return fallback()
+        }
+
+        let bin = String(bin.prefix(BinInfoProvider.minBinLength))
+        if let cached = cache[bin] {
+            completion(cached)
```

Comment:
should we have an early return here?
if yes - maybe guard is a good idea here, since this sounds like a precondition we check.

Reply:
right! Good one!